gslib/commands/iam.py

```diff
@@ -363,8 +374,153 @@ def get_gcloud_storage_args(self):
       # Gsutil takes policy file then URLs, and gcloud storage does the reverse.
       self.args = url_strings + self.args[:1]
 
+    elif self.sub_command == 'ch':
+      if shim_util.HIDDEN_SHIM_MODE(
+          config.get('GSUtil', 'hidden_shim_mode',
+                     'none')) is shim_util.HIDDEN_SHIM_MODE.DRY_RUN:
+        self.logger.warning(
+            'The shim maps iam ch commands to several gcloud storage commands,'
+            ' which cannot be determined without running gcloud storage.')
+      return []  # Translation occurs in self.run_gcloud_storage.
+
     return super().get_gcloud_storage_args(gcloud_storage_map)
 
+  def _RaiseIfInvalidUrl(self, url):
+    if url.IsFileUrl():
+      error_msg = 'Invalid Cloud URL "%s".' % url.object_name
+      if set(url.object_name).issubset(set('-Rrf')):
+        error_msg += (
+            ' This resource handle looks like a flag, which must appear'
+            ' before all bindings. See "gsutil help iam ch" for more details.')
+      raise CommandException(error_msg)
+
+  def _GetSettingsAndDiffs(self):
+    self.continue_on_error = False
+    self.recursion_requested = False
+
+    patch_bindings_tuples = []
+
+    if self.sub_opts:
+      for o, a in self.sub_opts:
+        if o in ['-r', '-R']:
+          self.recursion_requested = True
+        elif o == '-f':
+          self.continue_on_error = True
+        elif o == '-d':
+          patch_bindings_tuples.append(BindingStringToTuple(False, a))
+
+    url_pattern_strings = []
+
+    # N.B.: self.sub_opts stops taking in options at the first non-flagged
+    # token. The rest of the tokens are sent to self.args. Thus, in order to
+    # handle input of the form "-d <binding> <binding> <url>", we will have to
+    # parse self.args for a mix of both bindings and CloudUrls. We are not
+    # expecting to come across the -r, -f flags here.
+    it = iter(self.args)
+    for token in it:
+      if (STORAGE_URI_REGEX.match(token) and
+          IsKnownUrlScheme(GetSchemeFromUrlString(token))):
+        url_pattern_strings.append(token)
+        break
+      if token == '-d':
+        patch_bindings_tuples.append(BindingStringToTuple(False, next(it)))
```

Comment:
The empty iterator error message might be a bit confusing here

Reply:
Fixed.